Per connection local rate limiting

gokulnair opened the pull request:
Commit Message:
This is a PR for scoping token buckets in the local rate limiting flow on a per connection basis as opposed to scoping it on the entire envoy instance. More details in #15637 

Additional Description:
Currently, the HTTP local rate limiter's token bucket is shared across all workers, thus causing the rate limits to be applied per Envoy instance/process. This could potentially result in bad actors quickly exhausting limits on a given envoy instance before legitimate users have had a fair chance. We achieve this by adding an instance of the `LocalRateLimit::LocalRateLimiterImpl` to each connection object,  if there isn't one already, via `FilterState` data

Risk Level: Low

Testing:
Added unit tests to [local_ratelimit](https://github.com/envoyproxy/envoy/blob/main/test/extensions/filters/http/local_ratelimit/filter_test.cc)
Manually tested via curl'ing against a locally patched envoy instance.
One can send multiple requests on the same connection via curl using the following:
`
curl -vI example.com example.com
`

Docs Changes:
Added new toggle to local rate limit configuration to enable per connection local rate limiting

```
// Specifies the scope of the rate limiter's token bucket. 
// If set to false, the token bucket is shared across all worker threads
// thus the rate limits are applied per Envoy process.
// If set to true, a token bucket is allocated for each connection.
// Thus the rate limits are applied per connection thereby allowing
// one to rate limit requests on a per connection basis.
// If unspecified, the default value is false.
bool local_rate_limit_per_downstream_connection
```

Sample configuration
```name: envoy.filters.http.local_ratelimit
typed_config:
  "@type": type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
  stat_prefix: http_local_rate_limiter
  token_bucket:
    max_tokens: 10000
    tokens_per_fill: 1000
    fill_interval: 1s
  filter_enabled:
    runtime_key: local_rate_limit_enabled
    default_value:
      numerator: 100
      denominator: HUNDRED
  filter_enforced:
    runtime_key: local_rate_limit_enforced
    default_value:
      numerator: 100
      denominator: HUNDRED
  response_headers_to_add:
    - append: false
      header:
        key: x-local-rate-limit
        value: 'true'
  local_rate_limit_per_downstream_connection: true
```

[Optional Fixes #Issue]
#15637

gokulnair commented:
cc @mattklein123 @rgs1

rgs1 commented:
@gokulnair mind addressing the format error:

```
trailing whitespace: source/extensions/filters/http/local_ratelimit/local_ratelimit.h
```

I'll review the rest in the meanwhile, thanks!

rgs1 commented on api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto:
```diff
@@ -91,4 +91,6 @@ message LocalRateLimit {
   //  The filter supports a range of 0 - 10 inclusively for stage numbers.
   uint32 stage = 9 [(validate.rules).uint32 = {lte: 10}];
+
+  bool per_connection = 10;
```
Let's put a long description her explaining what this knob does, when it might be a good idea to use it, etc, etc.

gokulnair replied:
Would it make sense to make this name a little more descriptive, such as `rate_limiter_per_connection` maybe?

rgs1 replied:
Yes, I think that would help. I think `rate_limiter_per_connection` is good, given that we have `connection_pool_per_downstream_connection` in the cluster msg definition. Alternatively `apply_per_connection` is a suggestion if you are unconvinced.

rgs1 replied:
Alternatively, `per_connection_local_rate_limiter` is also good. I think I like that one the most.

rgs1 commented on source/extensions/filters/http/local_ratelimit/config.cc:
```diff
@@ -18,8 +18,10 @@ Http::FilterFactoryCb LocalRateLimitFilterConfig::createFilterFactoryFromProtoTy
-    callbacks.addStreamFilter(std::make_shared<Filter>(filter_config));
+  Event::Dispatcher& dispatcher = context.dispatcher();
+
+  return [filter_config, &dispatcher](Http::FilterChainFactoryCallbacks& callbacks) -> void {
```
`filter_config` takes the dispatcher as an argument to its constructer, why do we need to pass it here again? Let's just have filters access the dispatcher via `filter_config`.

gokulnair replied:
That's exactly what I had initially as well, i.e. accessing the dispatcher via the `filter_config`, only to realize that we are using a `const` pointer to the `filter_config` and as a result accessing non-const members/methods wouldn't work on it (dispatcher needs to be non-const).

rgs1 replied:
You can use `mutable`...

rgs1 replied:
oh you mean `getConfig()` here:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/local_ratelimit/local_ratelimit.h#L110

that's fine, it doesn't have to be const we can drop that....

rgs1 replied:
Actually, don't drop the `const` in `getConfig()` -- it's fine to return a non const reference from a const method:

https://godbolt.org/z/P6vvarhn3

gokulnair replied:
Turns out the event `dispatcher` from `FilterConfig` isn't thread safe for use from within the stream filter which executes in a different thread. So I switched to using the thread local `dispatcher` instead.

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -85,6 +96,7 @@ class FilterConfig : public Router::RouteSpecificFilterConfig {
   Router::HeaderParserPtr response_headers_parser_;
   const uint64_t stage_;
   const bool has_descriptors_;
+  const envoy::extensions::filters::http::local_ratelimit::v3::LocalRateLimit& proto_config_;
```
I think we need to copy the proto object, since this reference could (will) go stale.

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -106,9 +119,12 @@ class Filter : public Http::PassThroughFilter {
 
   const FilterConfig* getConfig() const;
   FilterConfigSharedPtr config_;
+  Event::Dispatcher& dispatcher_;
```
per comment above, not needed since we can get it via `config_`

rgs1 reviewed (changes requested):
Did a first pass, let's also add docs and tests. Thanks!

gokulnair commented:
> Did a first pass, let's also add docs and tests. Thanks!

Thanks for taking a look! I've moved it out of draft and added docs and tests as well. Let me know if this is good to go.

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:
```diff
@@ -106,6 +114,34 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
+          PerConnectionRateLimiter::key())) {
+    auto rate_limiter = std::make_shared<Filters::Common::LocalRateLimit::LocalRateLimiterImpl>(
+        std::chrono::milliseconds(
+            PROTOBUF_GET_MS_OR_DEFAULT(config->protoConfig().token_bucket(), fill_interval, 0)),
```
we can avoid doing this `std::chrono::milliseconds()` creation on every new connection by doing it once when constructing the FilterConfig object.

gokulnair replied:
This makes sense. I've made the appropriate amends 👍

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:
```diff
@@ -106,6 +114,34 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
+        std::chrono::milliseconds(
+            PROTOBUF_GET_MS_OR_DEFAULT(config->protoConfig().token_bucket(), fill_interval, 0)),
+        config->protoConfig().token_bucket().max_tokens(),
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(config->protoConfig().token_bucket(), tokens_per_fill, 1),
```
ditto

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -42,6 +42,19 @@ struct LocalRateLimitStats {
+  LocalRateLimiterImplSharedPtr value() const { return rate_limiter_; }
+
+private:
+  LocalRateLimiterImplSharedPtr rate_limiter_;
```
does this really need to be shared? Sounds like it could be unique or -- possibly better -- just an instance constructed when PerConnectionRateLimiter is created? `value()` would then just return a `LocalRateLimiterImpl&`.

gokulnair replied:
Seems like this can only be a `shared_ptr`. `PerConnectionRateLimiter` stores the state, i.e. an object of type `LocalRateLimiterImpl`,  inside it but since the latter does not have a copy constructor, you cannot pass in an already created instance nor can you create it within `PerConnectionRateLimiter` and return it since both of these would need a copy constructor. 
A `unique_ptr` wouldn't work for similar reasons since multiple, potentially concurrent, streams would be operating on a single instance of `LocalRateLimiterImpl` so transferring ownership of the unique_ptr out of `PerConnectionRateLimiter` and to each calling stream wouldn't make sense, if I'm understanding this correctly.

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -53,6 +66,9 @@ class FilterConfig : public Router::RouteSpecificFilterConfig {
   const LocalInfo::LocalInfo& localInfo() const { return local_info_; }
   Runtime::Loader& runtime() { return runtime_; }
+  const envoy::extensions::filters::http::local_ratelimit::v3::LocalRateLimit protoConfig() const {
+    return proto_config_;
```
`const envoy::extensions::filters::http::local_ratelimit::v3::LocalRateLimit&`? Do we need a full copy of the proto?

gokulnair replied:
I had this as a reference in the beginning but I believe you had pointed out that the reference could go stale:

> I think we need to copy the proto object, since this reference could (will) go stale.

Either way, I'm now storing the individual data we need under `FilterConfig` and have gotten rid of the full copy of the proto but now we're in a situation where we're individually storing more or less the same amount of data in `FilterConfig` maybe with a few exceptions ...

rgs1 reviewed (changes requested):
Nice progress, left some comments -- thanks!

gokulnair commented:
Ugh! I might've messed up this dev branch by fetching & rebasing from upstream and resolving conflicts rather than resolving it directly via github 🤦‍♂️Not sure if there's an easier way to restore this branch other than closing this PR and reopening a new one?

jmarantz commented:
FWIW I don't know of a situation where it's appropriate to rebase in the Envoy github flow. Just merge main and resolve conflicts that way.

phlax commented on api/BUILD:
```diff
@@ -133,12 +133,18 @@ proto_library(
         "//envoy/config/common/matcher/v3:pkg",
         "//envoy/config/core/v3:pkg",
         "//envoy/config/endpoint/v3:pkg",
+        "//envoy/config/filter/thrift/router/v2alpha1:pkg",
```
these shouldnt be here - i think they  have just been removed

rgs1 commented on docs/root/configuration/http/http_filters/local_rate_limit_filter.rst:
```diff
@@ -22,8 +22,9 @@ configured to be returned.
-  The token bucket is shared across all workers, thus the rate limits are applied per Envoy process.
+Depending on the value of the config :ref:`local_rate_limit_per_downstream_connection <envoy_v3_api_field_extensions.filters.http.local_ratelimit.v3.LocalRateLimit.local_rate_limit_per_downstream_connection>`,
+the token bucket is either shared across all workers or on a per connection basis. This results in the local rate limits being applied either per Envoy process or per downstream connection.
+If unspecified, this config is set to false by default, which means the rate limits are applied per Envoy process.
```
Probably simplify this last sentence into:

By default rate limits are applied per Envoy process.

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -62,6 +75,15 @@ class FilterConfig : public Router::RouteSpecificFilterConfig {
   Http::Code status() const { return status_; }
   uint64_t stage() const { return stage_; }
   bool hasDescriptors() const { return has_descriptors_; }
+  std::chrono::milliseconds fillInterval() const { return fill_interval_; }
```
nit return `const std::chrono::milliseconds&`

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -62,6 +75,15 @@ class FilterConfig : public Router::RouteSpecificFilterConfig {
+  uint32_t maxTokens() const { return max_tokens_; }
+  uint32_t tokensPerFill() const { return tokens_per_fill_; }
+  const Protobuf::RepeatedPtrField<
+      envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>
```
nit: 

```
 const Protobuf::RepeatedPtrField<
      envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>&
```

(a reference)

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.h:
```diff
@@ -108,6 +137,8 @@ class Filter : public Http::PassThroughFilter {
 
   void populateDescriptors(std::vector<RateLimit::LocalDescriptor>& descriptors,
                            Http::RequestHeaderMap& headers);
+  LocalRateLimiterImplSharedPtr getRateLimiter();
```
return `LocalRateLimiterImpl&`?

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:
```diff
@@ -84,7 +90,10 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   }
 
-  if (config->requestAllowed(descriptors)) {
+  bool isRequestAllowed = config->rateLimitPerConnection() ? requestAllowed(descriptors)
```
style: `const bool is_request_allowed  = ... `

rgs1 commented on source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:
```diff
@@ -109,6 +118,31 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
+}
+
+const Filters::Common::LocalRateLimit::LocalRateLimiterImpl& Filter::getRateLimiter() {
+  const auto* config = getConfig();
```
move this closer to where it's used, e.g.: just before `decoder_callbacks_->streamInfo().filterState()->setData()`

gokulnair replied:
Moved it back up a couple of lines to add an assertion.

rgs1 reviewed (commented):
LGTM, just two more nits. 

@antoniovicente any chance you could do a final review for this given that Matt is away?

rgs1 reviewed (dismissed):
LGTM generally, over to @htuch for API.

htuch reviewed (commented):
/lgtm api

gokulnair commented:
Thanks for the reviews! Let me know if there's anything else that's needed to get this going.

alyssawilk commented on api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto:
```diff
@@ -97,4 +97,12 @@ message LocalRateLimit {
   uint32 stage = 9 [(validate.rules).uint32 = {lte: 10}];
+
+  // Specifies the scope of the rate limiter's token bucket.
+  // If set to false, the token bucket is shared across all worker threads,
```
optional: it would be nice to make it clear here, as you did in docs, that false is the default.
(it's generally clear if you know proto but I'm not sure if everyone reading Envoy APIs knows proto defaults)

gokulnair replied:
Sounds good! I'll make it explicit as to what would be the default value if unspecified.

alyssawilk commented on docs/root/configuration/http/http_filters/local_rate_limit_filter.rst:
```diff
@@ -22,8 +22,9 @@ configured to be returned.
-  The token bucket is shared across all workers, thus the rate limits are applied per Envoy process.
+Depending on the value of the config :ref:`local_rate_limit_per_downstream_connection <envoy_v3_api_field_extensions.filters.http.local_ratelimit.v3.LocalRateLimit.local_rate_limit_per_downstream_connection>`,
+the token bucket is either shared across all workers or on a per connection basis. This results in the local rate limits being applied either per Envoy process or per downstream connection.
+By default the rate limits are applied per Envoy process.
```
It wasn't obviously clear to me why one would need this knob instead of just setting concurrent streams per connection.

Would stream limits work for your use case, or do you explicitly want to rate limit for some routes and not others?  If the latter is the case I think it'd be useful to doc up an example of how you could apply limits to routes A+B and not limit C, and how the cap would be applied.

rgs1 replied:
Mentioning this – which Alyssa pointed to offline – which might or might not be Gokul's case, but `max_concurrent_streams` is only for H2/H3.

So this would still be necessary for H1.

alyssawilk replied:
But isn't this a per-connection thing?  HTTP/1 can only have one stream per connection at a time since it's just not multiplexed :-)

gokulnair replied:
I believe it comes down to the fact that the per connection quotas set and enforced via the [token bucket configuration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket) doesn't really care if the requests were concurrent. The SETTINGS_MAX_CONCURRENT_STREAMS, from what I understand, allows one to set a hard limit on the maximum number of pending or outstanding requests on the connection. On the other hand, setting a per connection rate limit quota of 25 tokens with a 5 sec fill_interval, for instance, would enforce it over that time period even if the requests were made serially one after another. If this makes sense, I can go ahead and add a blurb to this doc explaining this difference 👍

alyssawilk reviewed (commented):
Oops, I'm sorry about review lag - Matt would usually take the lead here but his leave got unexpectedly extended.  I'll pick it up instead!

gokulnair commented:
> Oops, I'm sorry about review lag - Matt would usually take the lead here but his leave got unexpectedly extended. I'll pick it up instead!

Thanks @alyssawilk!

alyssawilk reviewed (commented):
Looks totally solid!  Just a few minor nits, and thanks for your patience both with review lag and me getting a handle on use case :-)

antoniovicente commented on test/extensions/filters/http/local_ratelimit/filter_test.cc:
```diff
@@ -136,16 +158,55 @@ TEST_F(FilterTest, RequestRateLimited) {
+            filter_->decodeHeaders(request_headers, false));
+  EXPECT_EQ(Http::FilterHeadersStatus::Continue, filter_2_->decodeHeaders(request_headers, false));
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_2_->decodeHeaders(request_headers, false));
```
A comment explaining the differences between this test and FilterTest.RequestRateLimited may be helpful for future readers.  I think this shows that the limit is applied by connection by verifying that filter_2_ allows requests after filter_ hits the rate limit.

gokulnair replied:
Added a brief explanation to the test.

antoniovicente reviewed (commented):
Looks good, just some minor comments / nits.

gokulnair commented on api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto:
```diff
@@ -97,4 +97,12 @@ message LocalRateLimit {
+  // If set to true, a token bucket is allocated for each connection.
+  // Thus the rate limits are applied per connection thereby allowing
+  // one to rate limit requests on a per connection basis.
+  bool local_rate_limit_per_downstream_connection = 11;
```
It's definitely one or the other as it stands currently as I'm not sure the added complexity of dealing with potentially conflicting token bucket quotas between the per process and per connection configurations and the precedence rules we'd have to handle buys us much in the way of functionality.

We did indeed consider an enum initially but it didn't seem like there were very many realistic use cases mainly because a lot of the toggles that were based on certain request characteristics such as IP, Cookie etc can be handled today via [request descriptors](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter) ...

gokulnair commented:
> Looks good, just some minor comments / nits.

Thanks!

gokulnair commented:
I've addressed all comments/nits. Will take a look at adding the integration test shortly.

gokulnair commented on test/extensions/filters/http/local_ratelimit/filter_test.cc:
```diff
@@ -136,16 +158,55 @@ TEST_F(FilterTest, RequestRateLimited) {
   EXPECT_EQ(1U, findCounter("test.http_local_rate_limit.rate_limited"));
 }
 
+TEST_F(FilterTest, RequestRateLimitedPerConnection) {
```
Sure, I can take a look at adding an integration test for the above scenario. The remaining comments should be addressed and good to go.

gokulnair replied:
Ah! After taking a closer look, seems like the tests in ratelimit_integration_test.cc are all targeting the 'rate limiter as an external service' use case and it makes complete sense to have integration tests for that.
For local rate limiting within envoy however, doesn't really look like we have any integration tests and I'm wondering if it buys us anything more than what the unit tests do currently. Thoughts? cc @rgs1

alyssawilk replied:
I think integration tests generally have value - I wasn't convinced setting connection data on the per-request StreamInfo would work as expected (there are no currently filters which do this, but I went ahead and tweaked some existing code to verify it worked for my own satiscation =P) but given the review delay this PR already suffered and lack of local rate limit tests to crib off of, I'm inclined to let it go this once though if you want to do a follow-up you'd totally earn brownie points :-)

gokulnair replied:
Brownie points it is then :-) I do have a draft up but I'm fighting the integration test api a bit at the moment, specifically with trying to get it to send multiple requests over the same connection while still ensuring that the first one generates an upstream request over its `fake_upstreams_` while the subsequent locally rate limited ones don't. It's holding it up more than necessary so a follow up task is making more sense at this point.

Thanks for taking a look!!

gokulnair commented on test/extensions/filters/http/local_ratelimit/filter_test.cc:
```diff
@@ -39,6 +39,7 @@ stat_prefix: test
     header:
       key: x-local-ratelimited
       value: 'true'
+local_rate_limit_per_downstream_connection: {}
```
Added comments below.

antoniovicente reviewed (approved):
Looks good.  I'm sure that the integration test you're looking into would make it even better.

Thanks!

gokulnair commented:
/restest

antoniovicente commented:
/retest

alyssawilk reviewed (approved):
LGTM

Could you please update the PR description with the appropriate fields  (testing: unit tests, docs: inline etc) and then I'll let @antoniovicente merge when he's set.

antoniovicente commented:
> LGTM
> 
> Could you please update the PR description with the appropriate fields (testing: unit tests, docs: inline etc) and then I'll let @antoniovicente merge when he's set.

I don't have remaining comments.  The remaining question is wherever or not integration tests are in the work as part of this PR or a followup.  A followup makes the most sense to me at this point, we should go ahead and merge.

gokulnair commented:
> LGTM
> 
> Could you please update the PR description with the appropriate fields (testing: unit tests, docs: inline etc) and then I'll let @antoniovicente merge when he's set.

gokulnair commented:
> I don't have remaining comments. The remaining question is wherever or not integration tests are in the work as part of this PR or a followup. A followup makes the most sense to me at this point, we should go ahead and merge.

Thanks! I do agree that a follow task makes the most sense at this time (I do have a draft up but its still WIP)
Also, I've updated the PR description with more details. Looks like we should be good to go for the merge, let me know if there's anything else that's needed from my end 👍

gokulnair commented:
> /lgtm api

@htuch Mind approving the api change once again, please? (we updated a doc in the proto file) 
Thanks!

htuch reviewed (commented):
/lgtm api
